File: utils/logger.go
package utils

import (
	"log/slog"
	"os"
)

var Logger *slog.Logger

func init() {
	// Initialize with a default text handler to stderr
	Logger = slog.New(slog.NewTextHandler(os.Stderr, &slog.HandlerOptions{
		Level: slog.LevelInfo,
	}))
}

func InitLogger(level slog.Level, jsonOutput bool) {
	opts := &slog.HandlerOptions{
		Level: level,
	}

	var handler slog.Handler
	if jsonOutput {
		handler = slog.NewJSONHandler(os.Stderr, opts)
	} else {
		handler = slog.NewTextHandler(os.Stderr, opts)
	}

	Logger = slog.New(handler)
}

func Debug(msg string, args ...any) {
	Logger.Debug(msg, args...)
}

func Info(msg string, args ...any) {
	Logger.Info(msg, args...)
}

func Warn(msg string, args ...any) {
	Logger.Warn(msg, args...)
}

func Error(msg string, args ...any) {
	Logger.Error(msg, args...)
}

File: backend/src/infrastructure/logger.ts
type Level = 'INFO' | 'WARN' | 'ERROR';

function fmt(level: Level, msg: string, meta?: Record<string, unknown>): string {
  const ts = new Date().toISOString();
  const base = `[${ts}] [${level}] ${msg}`;
  if (meta) {
    return `${base} ${JSON.stringify(meta)}`;
  }
  return base;
}

export const logger = {
  info(msg: string, meta?: Record<string, unknown>): void {
    console.log(fmt('INFO', msg, meta));
  },
  warn(msg: string, meta?: Record<string, unknown>): void {
    console.warn(fmt('WARN', msg, meta));
  },
  error(msg: string, meta?: Record<string, unknown>): void {
    console.error(fmt('ERROR', msg, meta));
  },
};
